pageserver/src/tenant/timeline.rs

```diff
@@ -4756,6 +4770,18 @@ impl Timeline {
     }
 }
 
+impl Drop for Timeline {
+    fn drop(&mut self) {
+        if let Some(ancestor) = &self.ancestor_timeline {
+            // This lock should never be poisoned, but in case it is we do a .map() instead of
+            // an unwrap(), to avoid panicking in a destructor and thereby aborting the process.
+            if let Ok(mut gc_info) = ancestor.gc_info.write() {
+                gc_info.remove_child(self.timeline_id)
+            }
+        }
+    }
+}
```

Comment:
Locking from destructors is generally deadlock-risky.
I think it's ok here because the ancestor pointers all go in one direction.

Reply:
Yeah, pretty sure this is safe.  But later we might add the ability for child shards to post updates to their parents about things like what key ranges they might read -- that will probably also be safe to do via a mutex, but becomes increasingly ugly & might justify adding some non-blocking mailbox type thing here.